### ISSUE TITLE
`zef info ...` does not always show the most current version known

### ISSUE THREAD

**Issue by szabgab** [ugexe/zef#172]
After a Rakudo Star 2017.04 installation on OSX from a dmg file I tried to upgrade zef from the GitHub repository. `zef` reported the install to be successful, but `zef info` still reports the old version.

## Context

#122

## Expected Behavior

I was hoping to see `zef:ver('0.1.16')` being reported after installation (upgrade).

## Actual Behavior

Before:

```
$ zef info zef
- Info for: zef
- Identity: zef:ver('0.1.9'):auth('github:ugexe')
- Recommended By: /Applications/Rakudo/share/perl6/site
Description:	 It's like [cpanm] wearing high heels with a tracksuit
License:	 Artistic-2.0
Source-url:	 git://github.com/ugexe/zef.git
Provides: 34 modules
Support:
#   bugtracker:	https://github.com/ugexe/zef/issues
Depends: 0 items
```

```
$ git clone https://github.com/ugexe/zef.git
$ cd zef
$ perl6 -Ilib bin/zef install -force-install .
===> Testing: zef:ver('0.1.16'):auth('github:ugexe')
===> Testing [OK] for zef:ver('0.1.16'):auth('github:ugexe')
===> Installing: zef:ver('0.1.16'):auth('github:ugexe')

1 bin/ script [zef] installed to:
/Applications/Rakudo/share/perl6/site/bin
```

After
```
$ zef info zef
- Info for: zef
- Identity: zef:ver('0.1.9'):auth('github:ugexe')
- Recommended By: /Applications/Rakudo/share/perl6/site
Description:	 It's like [cpanm] wearing high heels with a tracksuit
License:	 Artistic-2.0
Source-url:	 git://github.com/ugexe/zef.git
Provides: 34 modules
Support:
#   bugtracker:	https://github.com/ugexe/zef/issues
Depends: 0 items
```

```
$ which zef
/Applications/Rakudo/share/perl6/site/bin//zef
```

## Steps to Reproduce
```
$ perl6 -Ilib bin/zef --debug install -force-install .
The following local paths candidates are already installed: .
===> Filtering: zef:ver('0.1.16'):auth('github:ugexe')
===> Filtering [OK] for zef:ver('0.1.16'):auth('github:ugexe')
===> # SKIP: No Build.pm for zef:ver('0.1.16'):auth('github:ugexe')
===> Testing: zef:ver('0.1.16'):auth('github:ugexe')
Testing with plugin: Zef::Service::TAP+{<anon|140457825883952>}
t/00-load.t ........... ok
t/identity.t .......... ok
t/utils-filesystem.t .. ok
All tests successful.
Files=3, Tests=14,  2 wallclock secs
Result: PASS
===> Testing [OK] for zef:ver('0.1.16'):auth('github:ugexe')
===> Installing: zef:ver('0.1.16'):auth('github:ugexe')
===> Install [OK] for zef:ver('0.1.16'):auth('github:ugexe')

1 bin/ script [zef] installed to:
/Applications/Rakudo/share/perl6/site/bin
```

```
$ zef --debug info zef
- Info for: zef
- Identity: zef:ver('0.1.9'):auth('github:ugexe')
- Recommended By: /Applications/Rakudo/share/perl6/site
Description:	 It's like [cpanm] wearing high heels with a tracksuit
License:	 Artistic-2.0
Source-url:	 git://github.com/ugexe/zef.git
Provides: 34 modules
------------------------------------------------------------------------------------------------------------------------------------
Module                                    |Path-Name                                       |File-ID
------------------------------------------------------------------------------------------------------------------------------------
Zef                                       |lib/Zef.pm6                                     |0258AAFE512ECBC741A29D139154EDFB491C9388
Zef::CLI                                  |lib/Zef/CLI.pm6                                 |09969B10D9F270B8DBAAB9961FA7E6CDB8AD58C9
Zef::Test                                 |lib/Zef/Test.pm6                                |E5541742FCBBD3BB5ADFD4A8496A6BEDFE7A8919
Zef::Shell                                |lib/Zef/Shell.pm6                               |71AEF738564DCC3EE14AAEF53360609BE7F9DE61
Zef::Build                                |lib/Zef/Build.pm6                               |94973B3C3F2195BE0D924435ACA532C2B1695192
Zef::Fetch                                |lib/Zef/Fetch.pm6                               |A48CF21F42885E27DBC7EBC4927226101DEAD45B
Zef::Config                               |lib/Zef/Config.pm6                              |84962362F2FC3E01903C8DB77A5C66E7B282DF00
Zef::Client                               |lib/Zef/Client.pm6                              |4B1D5A60B59D9541E13F76E0E2A2D550E0144053
Zef::Extract                              |lib/Zef/Extract.pm6                             |26BEE6DE3D8587782D7BFBCA96F730387710091D
Zef::Identity                             |lib/Zef/Identity.pm6                            |5F1D4D8BA35C07FC729E459B8DB34F0331D9A991
Zef::Repository                           |lib/Zef/Repository.pm6                          |55E3A09C837FDC44EF621FC4F37162E35ABB468E
Zef::Utils::URI                           |lib/Zef/Utils/URI.pm6                           |8625527A2B0B5EA7E5391E97F8B760021F100261
Zef::Service::TAP                         |lib/Zef/Service/TAP.pm6                         |827E2DD47D5C1E4FB3D67DEB45D23A350171A8D0
Zef::Distribution                         |lib/Zef/Distribution.pm6                        |B4C3CCD80C5003A2A09E0891E9058A7E10D28768
Zef::Utils::FileSystem                    |lib/Zef/Utils/FileSystem.pm6                    |147FE99329D7548EEAEC9BB1952F90BE1E972DA8
Zef::Utils::SystemInfo                    |lib/Zef/Utils/SystemInfo.pm6                    |A105E3FD72B9B05F0D1476B6FC87E830809D041C
Zef::Service::FetchPath                   |lib/Zef/Service/FetchPath.pm6                   |33246461AB4FA8D26ED3B3F32FAA83C7EA285EC5
Zef::Distribution::Local                  |lib/Zef/Distribution/Local.pm6                  |D2D7BE27C4415C0CA56E7907E18AF2E002594255
Zef::Service::Shell::tar                  |lib/Zef/Service/Shell/tar.pm6                   |5E8059ADE560D52824902349953178A448C97F3F
Zef::Service::Shell::git                  |lib/Zef/Service/Shell/git.pm6                   |8DB555D333775007962C71E559E8ABE0B0DF3FFD
Zef::Repository::MetaCPAN                 |lib/Zef/Repository/MetaCPAN.pm6                 |CA5564B48E997425282CA41539598D9695C42FB0
Zef::Service::Shell::curl                 |lib/Zef/Service/Shell/curl.pm6                  |9C81E827621972AD941406C100983436EA920FCF
Zef::Service::Shell::wget                 |lib/Zef/Service/Shell/wget.pm6                  |5A4B5C340865C9A05C2268664E2E7F3B720FF3E1
Zef::Service::Shell::Test                 |lib/Zef/Service/Shell/Test.pm6                  |8C6561A7C1308FF5C78349A27DB702C92BFB606A
Zef::Service::Shell::unzip                |lib/Zef/Service/Shell/unzip.pm6                 |27CCC59AF8016FA56B0E6A7B79AA45DE59737AF9
Zef::Service::Shell::prove                |lib/Zef/Service/Shell/prove.pm6                 |B51F88EA66B0A94EA08DA2557EE2DEC8A72A5A89
Zef::Service::Shell::p5tar                |lib/Zef/Service/Shell/p5tar.pm6                 |BCCBC804F562B2B4C49173F5241C79888BCE4BAA
Zef::Service::Shell::Build                |lib/Zef/Service/Shell/Build.pm6                 |7306A354D0C1F3E156AF88286C0876E946703595
Zef::Repository::LocalCache               |lib/Zef/Repository/LocalCache.pm6               |F4EED3DB31B17000B7877FE3B0412D4F273133C7
Zef::Repository::Ecosystems               |lib/Zef/Repository/Ecosystems.pm6               |647FBAA7730CF893D4ED40D1EDB292292DB0F3BF
Zef::Service::Shell::PowerShell           |lib/Zef/Service/Shell/PowerShell.pm6            |6F415DD417565B7D1EAA9D07E9AD301AD572B825
Zef::Service::Shell::PowerShell::unzip    |lib/Zef/Service/Shell/PowerShell/unzip.pm6      |AC2FD338DEF51F990563C6291A4176FE7A11299E
Zef::Service::Shell::PowerShell::download |lib/Zef/Service/Shell/PowerShell/download.pm6   |DD1732CDF3A6F0235114E16397742CA76DCA3F1E
Zef::Distribution::DependencySpecification|lib/Zef/Distribution/DependencySpecification.pm6|804BBC43C630E9E5A9E8AB3F34F16BB152857FD3
------------------------------------------------------------------------------------------------------------------------------------
Support:
#   bugtracker:	https://github.com/ugexe/zef/issues
Depends: 0 items
```
## Your Environment

```
$ perl6 -v
This is Rakudo version 2017.04.3 built on MoarVM version 2017.04-53-g66c6dda
implementing Perl 6.c.
```

```
$ perl6 -e 'say $*VM.config<os>'
darwin
```

**Comment by ugexe** [comment]
I've already tried to reproduce your issue with no luck - nor did you respond to it.
https://github.com/ugexe/zef/issues/168

Since I reproduced all the steps in question to UPGRADE (not force install as you are doing) and it worked in the link above.

Can you please be more careful and detailed before sending more bug reports? This is like 5 bugs now that are not valid or acceptable.

**Comment by szabgab** [comment]
You asked me to provide the details by filling out the form. I just did it. Though I see you asked to reopen the ticket and I opened a new one.

**Comment by ugexe** [comment]
You are not upgrading in this ticket (although you were in the previous ticket) but in the title you say zef fails to upgrade. I mentioned in the last ticket that I could not reproduce with newer versions and posted my attempt at reproducing your issue - except my attempt has the expected results. I have now, in this ticket, tried to reproduce and install of multiple copies of zef - but still get the expected results.

**Comment by szabgab** [comment]
So there might be some issue with my system. I was hoping you'll give me some extra command or options, like the `--debug`, that will provide enough information to shed light what is the issue, that apparently only happens on my system.

I tried now the following in the zef clone:
```
$ perl6 -Ilib bin/zef --debug upgrade .
Can't upgrade identities that aren't installed: .
```
which in itself is surprising me as I've just force-installed it from the same directory.

then still in the zef clone I tried:
```
$ perl6 -Ilib bin/zef --debug upgrade zef
===> Searching for: zef
===> Found: zef:ver('0.1.16'):auth('github:ugexe') [via Zef::Repository::Ecosystems<p6c>]
The following distributions are already at their latest versions: zef:ver('0.1.16'):auth('github:ugexe')
```

then
```
$ zef --debug upgrade zef
===> Searching for: zef
===> Found: zef:ver('0.1.16'):auth('github:ugexe') [via Zef::Repository::Ecosystems<p6c>]
The following distributions are already at their latest versions: zef:ver('0.1.16'):auth('github:ugexe')
```

I still get
```
$ zef info zef
- Info for: zef
- Identity: zef:ver('0.1.9'):auth('github:ugexe')
- Recommended By: /Applications/Rakudo/share/perl6/site
Description:	 It's like [cpanm] wearing high heels with a tracksuit
License:	 Artistic-2.0
Source-url:	 git://github.com/ugexe/zef.git
Provides: 34 modules
Support:
#   bugtracker:	https://github.com/ugexe/zef/issues
Depends: 0 items
```

Then following what I saw in your example:
```
$ perl6 -Ilib bin/zef --debug install "zef:ver<0.1.15>"
===> Searching for: zef:ver('0.1.15')
===> Found: zef:ver('0.1.15'):auth('github:ugexe') [via Zef::Repository::Ecosystems<cpan>]
===> Fetching: zef:ver('0.1.15')
Fetching with plugin: Zef::Service::Shell::wget+{<anon|140457825883952>}
===> Fetching [OK]: zef:ver('0.1.15'):auth('github:ugexe') to /Users/gabor/.zef/tmp/zef-0.1.15.tar.gz
===> Extracting: zef:ver('0.1.15')
Extracting with plugin: Zef::Service::Shell::tar+{<anon|140457825883952>}
===> Extraction [OK]: zef:ver('0.1.15') to /Users/gabor/.zef/store/zef-0.1.15.tar.gz
===> Filtering: zef:ver('0.1.15'):auth('github:ugexe')
===> Filtering [OK] for zef:ver('0.1.15'):auth('github:ugexe')
===> # SKIP: No Build.pm for zef:ver('0.1.15'):auth('github:ugexe')
===> Testing: zef:ver('0.1.15'):auth('github:ugexe')
Testing with plugin: Zef::Service::TAP+{<anon|140457825883952>}
t/00-load.t ........... ok
t/identity.t .......... ok
t/utils-filesystem.t .. ok
All tests successful.
Files=3, Tests=14,  21 wallclock secs
Result: PASS
===> Testing [OK] for zef:ver('0.1.15'):auth('github:ugexe')
===> Installing: zef:ver('0.1.15'):auth('github:ugexe')
===> Install [OK] for zef:ver('0.1.15'):auth('github:ugexe')

1 bin/ script [zef] installed to:
/Applications/Rakudo/share/perl6/site/bin
```
and I still get:
```
$ zef info zef
- Info for: zef
- Identity: zef:ver('0.1.9'):auth('github:ugexe')
- Recommended By: /Applications/Rakudo/share/perl6/site
Description:	 It's like [cpanm] wearing high heels with a tracksuit
License:	 Artistic-2.0
Source-url:	 git://github.com/ugexe/zef.git
Provides: 34 modules
Support:
#   bugtracker:	https://github.com/ugexe/zef/issues
Depends: 0 items
```

Is it possible that PATH I have reaches the old zef and not the one that is being installed?
This does not seem to be the case, but I think that zef file is actually just a launcher and not the real zef code.
```
$ which zef
/Applications/Rakudo/share/perl6/site/bin//zef
$ ls -l /Applications/Rakudo/share/perl6/site/bin//zef
-rwxr-xr-x  1 gabor  admin  1489 Jun 10 17:49 /Applications/Rakudo/share/perl6/site/bin//zef
$ date
Sat Jun 10 17:55:46 IDT 2017
```

**Comment by ugexe** [comment]
`perl6 -Ilib bin/zef --debug upgrade .` fails because upgrade does not take a path - you tell it *what* you want to upgrade, not *with what* you want to upgrade. This could probably be changed.

You can see which versions are installed manually by doing `perl6 -e 'use Zef:ver<0.1.15>;'` `perl6 -e 'use Zef:ver<0.1.16>;' etc

Finally you may wish to try the following to avoid using the rakudo bin/ script wrapper altogether:
1: `zef uninstall zef` # uninstalls all versions of zef
2: `perl6 -Ilib bin/zef --debug install "zef:ver<0.1.15>"`
3: `perl6 -Ilib bin/zef --debug install "zef:ver<0.1.16>"`
4: `perl6 -Ilib bin/zef info zef`

**Comment by ugexe** [comment]
I still cannot reproduce, but someone else has so this issue is confirmed. I believe I understand what the issue is, so will patch later tonight.

**Comment by ugexe** [comment]
This should be resolved in `zef:ver<0.1.17>` https://github.com/ugexe/zef/commit/f9289c74dabfdd3d7c995c33cac42adb7faf41cf

**Comment by szabgab** [comment]
After updating the the git repo it worked! Thank you!  🍺  🍪 


```
$ perl6 -Ilib bin/zef uninstall zef
===> Uninstalled from /Applications/Rakudo/share/perl6/site
zef:ver('0.1.9'):auth('github:ugexe')
zef:ver('0.1.16'):auth('github:ugexe')
zef:ver('0.1.15'):auth('github:ugexe')
```

```
$ perl6 -Ilib bin/zef --debug install zef
===> Searching for: zef
===> Found: zef:ver('0.1.17'):auth('github:ugexe') [via Zef::Repository::Ecosystems<p6c>]
===> Fetching: zef
Fetching with plugin: Zef::Service::Shell::git+{<anon|140457825883952>}
===> Fetching [OK]: zef:ver('0.1.17'):auth('github:ugexe') to /Users/gabor/.zef/tmp/zef.git
===> Extracting: zef
Extracting with plugin: Zef::Service::Shell::git+{<anon|140457825883952>}
===> Extraction [OK]: zef to /Users/gabor/.zef/store/zef.git
===> Filtering: zef:ver('0.1.17'):auth('github:ugexe')
===> Filtering [OK] for zef:ver('0.1.17'):auth('github:ugexe')
===> # SKIP: No Build.pm for zef:ver('0.1.17'):auth('github:ugexe')
===> Testing: zef:ver('0.1.17'):auth('github:ugexe')
Testing with plugin: Zef::Service::TAP+{<anon|140457825883952>}
t/00-load.t ........... ok
t/identity.t .......... ok
t/utils-filesystem.t .. ok
All tests successful.
Files=3, Tests=14,  21 wallclock secs
Result: PASS
===> Testing [OK] for zef:ver('0.1.17'):auth('github:ugexe')
===> Installing: zef:ver('0.1.17'):auth('github:ugexe')
===> Install [OK] for zef:ver('0.1.17'):auth('github:ugexe')

1 bin/ script [zef] installed to:
/Applications/Rakudo/share/perl6/site/bin
```

```
$ zef info zef
- Info for: zef
- Identity: zef:ver('0.1.17'):auth('github:ugexe')
- Recommended By: /Applications/Rakudo/share/perl6/site
Description:	 It's like [cpanm] wearing high heels with a tracksuit
License:	 Artistic-2.0
Source-url:	 git://github.com/ugexe/zef.git
Provides: 36 modules
Support:
#   bugtracker:	https://github.com/ugexe/zef/issues
Depends: 0 items
```
